### DISCUSSION TITLE
Wezterm support for variable weights.

### DISCUSSION THREAD

**Discussion by mindentropy** [wezterm/wezterm#6514] [General]
Does Wezterm support variable weights? I am using Wezterm version 20240203-110809-5046fc22 on Ubuntu and trying to use JetBrainsMono[wght].ttf font. I am setting the variable font weight but nothing seems to change. Am I missing something?

**Comment by habahabahaba** [comment]
I'm on Sonoma, but here are the settings, that worked for me:
```
-- FONTS:
config.font = wezterm.font_with_fallback {
  'GeistMono Nerd Font',
  'JetBrainsMono Nerd Font Mono',
  'Lilex Nerd Font Mono',
  'MesloLGS Nerd Font Mono', 
  'Noto Color Emoji',
}

-- For Geist:
config.line_height = 1.03
config.font_rules = { 
  -- Bold-not-italic 
  {
      intensity = 'Bold',
      italic = false,
      font = wezterm.font_with_fallback{
        { 
          family = 'GeistMono Nerd Font',
          weight = 1000,
        },
      }
  }, 
  -- Bold-and-italic
  {
      intensity = 'Bold',
      italic = true,
      font = wezterm.font_with_fallback {
        { 
          family = 'GeistMono Nerd Font',
          weight = 'ExtraBold',
          italic = true
        }
      }
  }, 
  -- -- normal-intensity-and-not-italic
  {
      intensity = 'Normal',
      italic = false,
      font = wezterm.font_with_fallback {
        { 
          family = 'GeistMono Nerd Font',
          weight = 500,
        }
      }
  }, 
  -- half-intensity-and-italic (half-bright or dim)
  {
      intensity = 'Half',
      italic = true,
      font = wezterm.font_with_fallback {
        { 
          family = 'GeistMono Nerd Font',
          weight = 'Light',
          italic = true
        }
      }
  }, 
  -- half-intensity-and-not-italic
  {
      intensity = 'Half',
      italic = false,
      font = wezterm.font_with_fallback {
        { 
          family = 'GeistMono Nerd Font',
          weight = 'Light'
        }
      }
  }
}
```

**Comment by mindentropy** [comment]
@habahabahaba : I have externally downloaded "JetBrains Mono" and not using the internal one. My simple config is as follows:

```
config.font = wezterm.font("JetBrains Mono", {weight = 500})
config.font_size = 21
config.font_rules = {
  {
	intensity = 'Bold',
	italic = false,

	font = wezterm.font {
		family = "JetBrains Mono",
--		family = "Recursive Monospace",
--		family = "Inconsolata",
		weight = 500,
--		stretch = "Expanded"
    }
  },

  {
	intensity = 'Bold',
	italic = true,

	font = wezterm.font {
		family = "JetBrains Mono",
--		family = "Iosevka Term SS07",
--		family = "Recursive Monospace",
--		family = "Inconsolata",
		weight = 500,
		style = "Italic",
--		stretch = "Expanded"
    }
  },
}
```
If I change the weight to 499, I get thin font. 500 suddenly becomes thick. My understanding of variable weight is that there is a gradual increase of the weights and not drastic increase.

The below image is on 500.

![image](https://github.com/user-attachments/assets/3f2b025d-bc57-437b-bfc4-28fc95b553e1)

The below image is on 499.

![image](https://github.com/user-attachments/assets/b0d6f55a-4281-44c7-95f9-c474201bdbfb)

**Reply by habahabahaba** [reply]
My GUESS would be: there's no 499 variant, so it defaults to something else. The weights are variable only if variants are provided.
Also, I have `italic = true,` and you - `style = "Italic",`

**Reply by mindentropy** [reply]
@habahabahaba : 

> My GUESS would be: there's no 499 variant, so it defaults to something else.

Shouldn't variable weight have a weight range? 

> Also, I have italic = true, and you - style = "Italic",

The rules what I have written is mentioned here -> https://wezfurlong.org/wezterm/config/lua/config/font_rules.html

**Reply by habahabahaba** [reply]
1. Unless the fonts get synthesized (which I'm not sure Wezterm can do), inside the range we can only use weights explicitely provided by the font.
2. Well, in your link: `
true (italic) or false (not italic)
`

**Reply by mindentropy** [reply]
> 1. Unless the fonts get synthesized (which I'm not sure Wezterm can do), inside the range we can only use weights explicitely provided by the font.


Sorry, I didn't get this. My ls-fonts shows
`.local/share/fonts/JetBrains/variable/JetBrainsMono[wght].ttf`

What do you mean by "weights explicitly provided by the font"?  Doesn't the variable font provide these? How can I check what weights a font provides?

**Reply by habahabahaba** [reply]
I mean, for every character of every style (italic or normal) and of every weight, this specific character must exist inside your font file. If your font-file has a capital letter "G" of weight 499 the Wezterm will display it, and if such character is not there, Wezterm will fall back on something else.

**Reply by mindentropy** [reply]
@habahabahaba : I have JetBrains Mono as external font. There is built in Wezterm font too. How does Wezterm select between these two just based on the name?

**Comment by codeRSH** [comment]
OK, I am also having the same issue. I am using Ubuntu Sans Mono Variable font. 

I just don't see any difference while using it with weight 400-499, but it becomes 'medium' when I use the weight = 500. 

Do note that I also run Ghostty, and the same font does change its width depending on the number specified. So there seems to be some issue with Wezterm I believe. 

Here is my font related config: 

```lua
  config.font = wezterm.font("Ubuntu Sans Mono", { weight = 450 })
  config.font_size = 14.0
  config.font_rules = {
    {
      intensity = "Bold",
      italic = false,
      font = wezterm.font("Ubuntu Sans Mono", { weight = 700 }),
    },
    {
      intensity = "Normal",
      italic = true,
      font = wezterm.font("Ubuntu Sans Mono", { weight = 450, italic = true }),
    },
  }
```

Should this be logged as an 'issue' or am I missing somthing?

**Reply by mindentropy** [reply]
I kind of partly solved it by experimenting with the configs. Following is what I have in Wezterm config and you might want to experiment with the combinations to get what you want.

```
--[[
For font config see the discussion comment below:
https://github.com/wezterm/wezterm/discussions/3388#discussioncomment-5446306
]]
config.freetype_load_flags = "NO_HINTING"
config.freetype_load_target = "Light"
config.freetype_render_target = "Normal"
-- config.freetype_render_target = "HorizontalLcd"
-- config.front_end = "WebGpu"
config.front_end = "OpenGL"

```
Support for variable font weights might be needed in Wezterm I think and this might be an issue I guess.
